src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsregisterstonomisupdate/services/CourtRegisterUpdateService.kt

```diff
@@ -1,17 +1,399 @@
+        if (currentCourtDataInPrisonSystem != null) translateToSync(currentCourtDataInPrisonSystem) else null
+      val newCourtData = mergeIds(updatedCourtDataFromRegister, currentCourtDataToCompare)
+
+      val diffs = checkForDifferences(currentCourtDataToCompare, newCourtData)
+      if (!diffs.areEqual()) {
+        log.info("$courtId: Updating Prison System with court data. Changes {}", diffs)
+
+        storeInPrisonData(currentCourtDataToCompare, newCourtData, applyChanges)
+        val trackingAttributes = mapOf(
+          "courtId" to courtId,
+          "changes" to diffs.toString(),
+          "changes-applied" to applyChanges.toString()
+        )
+        telemetryClient.trackEvent("HR2NU-Court-Change", trackingAttributes, null)
+      } else {
+        log.info("$courtId: No changes to apply")
+        telemetryClient.trackEvent("HR2NU-Court-No-Change", mapOf("courtId" to courtId), null)
+      }
+      return diffs
+    }
+    return null
+  }
+
+  private fun translateToSync(courtData: CourtFromPrisonSystem) =
+    CourtDataToSync(
+      courtData.agencyId,
+      courtData.description,
+      courtData.longDescription,
+      courtData.active,
+      null,
+      courtData.addresses.map { address ->
+        AddressDataToSync(
+          addressId = address.addressId,
+          addressType = getRefCode("ADDR_TYPE", address.addressType),
+          premise = address.premise,
+          street = address.street,
+          locality = address.locality,
+          town = getRefCode("CITY", address.town),
+          postalCode = address.postalCode,
+          county = getRefCode("COUNTY", address.county),
+          country = getRefCode("COUNTRY", address.country),
+          primary = address.primary,
+          noFixedAddress = address.noFixedAddress,
+          startDate = address.startDate,
+          endDate = address.endDate,
+          comment = address.comment,
+          phones = address.phones.map { phone ->
+            PhoneFromPrisonSystem(phone.phoneId, phone.number, phone.type, phone.ext)
+          }
+        )
+      }
+    )
+
+  private fun storeInPrisonData(
+    currentCourtData: CourtDataToSync?,
+    newCourtData: CourtDataToSync,
+    applyChanges: Boolean = false
+  ) {
+    if (applyChanges) {
+      val dataPayload = translateToPrisonSystemFormat(newCourtData)
+      if (currentCourtData == null) {
+        log.debug("Insert Court {}", dataPayload)
+        prisonService.insertCourt(dataPayload)
+      } else {
+        if (newCourtData != currentCourtData) { // don't update if equal
+          log.debug("Update Court {}", dataPayload)
+          prisonService.updateCourt(dataPayload)
+        }
+      }
+    }
+
+    currentCourtData?.addresses?.forEach {
+      if (newCourtData.addresses.find { a -> a.addressId == it.addressId } == null) {
+        log.info("No match found remove address {}", it)
+        if (applyChanges) prisonService.removeAddress(newCourtData.courtId, it.addressId!!)
+      }
```

Comment:
This might be something we go back to, Whether we decide to never deleted addresses. Although it would be good to know how we decided that a NOMIS address is the same as the COURT register address. Or do we think each time we update a court what will probably happen is the old address in NOMIS will be deleted and the new "similar " address would be added.